### ISSUE TITLE
double job run or get error

### ISSUE THREAD

**Issue by xurwxj** [vsivsi/meteor-file-job-sample-app#2]
when run meteor in bundle mode(production mode), image will generate two processing thumb after uploaded, the log as following:

```
App 402 stdout: Tue May 12 2015 20:00:59 GMT+0800 (CST), [SERVER], jobSave, params: [{"runId":null,"type":"makeThumb","data":{"owner":"RwkKCoX6L6tzmKDNj","inputFileURL":"http://test.3dker.com//gridfs/files/1217bc0772ae1006a7d9e39a","outputFileURL":"http://test.3dker.com//gridfs/files/put/6ceb1756bea6f39421a1744d","inputFileId":{"_str":"1217bc0772ae1006a7d9e39a"},"outputFileId":{"_str":"6ceb1756bea6f39421a1744d"}},"status":"waiting","updated":"2015-05-12T12:00:59.313Z","created":"2015-05-12T12:00:59.313Z","priority":0,"retries":6,"retryWait":20000,"retried":0,"retryBackoff":"constant","retryUntil":"+275760-09-13T00:00:00.000Z","repeats":0,"repeatWait":300000,"repeated":0,"repeatUntil":"+275760-09-13T00:00:00.000Z","after":"2015-05-12T12:01:04.316Z","progress":{"completed":0,"total":1,"percent":0},"depends":[],"resolved":[],"log":[{"time":"2015-05-12T12:00:59.316Z","runId":null,"level":"info","message":"Constructed"}]},{}]
App 357 stdout: Tue May 12 2015 20:00:59 GMT+0800 (CST), [SERVER], jobSave, params: [{"runId":null,"type":"makeThumb","data":{"owner":"RwkKCoX6L6tzmKDNj","inputFileURL":"http://test.3dker.com//gridfs/files/1217bc0772ae1006a7d9e39a","outputFileURL":"http://test.3dker.com//gridfs/files/put/8359016a2051b0941ed522de","inputFileId":{"_str":"1217bc0772ae1006a7d9e39a"},"outputFileId":{"_str":"8359016a2051b0941ed522de"}},"status":"waiting","updated":"2015-05-12T12:00:59.322Z","created":"2015-05-12T12:00:59.322Z","priority":0,"retries":6,"retryWait":20000,"retried":0,"retryBackoff":"constant","retryUntil":"+275760-09-13T00:00:00.000Z","repeats":0,"repeatWait":300000,"repeated":0,"repeatUntil":"+275760-09-13T00:00:00.000Z","after":"2015-05-12T12:01:04.323Z","progress":{"completed":0,"total":1,"percent":0},"depends":[],"resolved":[],"log":[{"time":"2015-05-12T12:00:59.323Z","runId":null,"level":"info","message":"Constructed"}]},{}]
App 402 stdout: Tue May 12 2015 20:00:59 GMT+0800 (CST), [SERVER], jobSave, returned: {"_str":"a7edab1a06bac3c3b16e4e7e"}
App 357 stdout: Tue May 12 2015 20:00:59 GMT+0800 (CST), [SERVER], jobSave, returned: {"_str":"29c283cb0e99b931f596ec05"}
App 402 stdout: Tue May 12 2015 20:01:05 GMT+0800 (CST), [SERVER], getWork, params: [["makeThumb"],{"maxJobs":52}]
App 402 stdout: Tue May 12 2015 20:01:05 GMT+0800 (CST), [SERVER], getWork, returned: [{"_id":{"_str":"29c283cb0e99b931f596ec05"},"runId":{"_str":"c501f04ff32209410d1914aa"},"type":"makeThumb","data":{"owner":"RwkKCoX6L6tzmKDNj","inputFileURL":"http://test.3dker.com//gridfs/files/1217bc0772ae1006a7d9e39a","outputFileURL":"http://test.3dker.com//gridfs/files/put/8359016a2051b0941ed522de","inputFileId":{"_str":"1217bc0772ae1006a7d9e39a"},"outputFileId":{"_str":"8359016a2051b0941ed522de"}},"status":"running","updated":"2015-05-12T12:01:05.468Z","created":"2015-05-12T12:00:59.326Z","priority":0,"retries":5,"retryWait":20000,"retried":1,"retryBackoff":"constant","retryUntil":"+275760-09-13T00:00:00.000Z","repeats":0,"repeatWait":300000,"repeated":0,"repeatUntil":"+275760-09-13T00:00:00.000Z","after":"2015-05-12T12:01:04.323Z","progress":{"completed":0,"total":1,"percent":0},"depends":[],"resolved":[]},{"_id":{"_str":"a7edab1a06bac3c3b16e4e7e"},"runId":{"_str":"c501f04ff32209410d1914aa"},"type":"makeThumb","data":{"owner":"RwkKCoX6L6tzmKDNj","inputFileURL":"http://test.3dker.com//gridfs/files/1217bc0772ae1006a7d9e39a","outputFileURL":"http://test.3dker.com//gridfs/files/put/6ceb1756bea6f39421a1744d","inputFileId":{"_str":"1217bc0772ae1006a7d9e39a"},"outputFileId":{"_str":"6ceb1756bea6f39421a1744d"}},"status":"running","updated":"2015-05-12T12:01:05.468Z","created":"2015-05-12T12:00:59.321Z","priority":0,"retries":5,"retryWait":20000,"retried":1,"retryBackoff":"constant","retryUntil":"+275760-09-13T00:00:00.000Z","repeats":0,"repeatWait":300000,"repeated":0,"repeatUntil":"+275760-09-13T00:00:00.000Z","after":"2015-05-12T12:01:04.316Z","progress":{"completed":0,"total":1,"percent":0},"depends":[],"resolved":[]}]
App 357 stdout: Tue May 12 2015 20:01:05 GMT+0800 (CST), [SERVER], getWork, params: [["makeThumb"],{"maxJobs":52}]
App 357 stdout: Tue May 12 2015 20:01:05 GMT+0800 (CST), [SERVER], getWork, returned: []
App 402 stdout: Tue May 12 2015 20:01:05 GMT+0800 (CST), [SERVER], jobProgress, params: [{"_str":"a7edab1a06bac3c3b16e4e7e"},{"_str":"c501f04ff32209410d1914aa"},20,100,{}]
App 402 stdout: Tue May 12 2015 20:01:05 GMT+0800 (CST), [SERVER], jobProgress, params: [{"_str":"29c283cb0e99b931f596ec05"},{"_str":"c501f04ff32209410d1914aa"},20,100,{}]
App 402 stdout: Tue May 12 2015 20:01:05 GMT+0800 (CST), [SERVER], jobProgress, returned: true
App 402 stdout: Tue May 12 2015 20:01:05 GMT+0800 (CST), [SERVER], jobProgress, returned: true
App 402 stdout: Tue May 12 2015 20:01:05 GMT+0800 (CST), [SERVER], jobFail, params: [{"_str":"29c283cb0e99b931f596ec05"},{"_str":"c501f04ff32209410d1914aa"},{},{"fatal":false}]
App 402 stderr:
App 402 stderr: /home/shining3d/3dker/programs/server/packages/meteor.js:1055
App 402 stderr:     Fiber(runWithEnvironment).run();
App 402 stderr:                               ^
App 402 stderr: Error: Match error: Expected plain object
App 402 stderr:     at checkSubtree (packages/check/match.js:279:1)
App 402 stderr:     at check (packages/check/match.js:32:1)
App 402 stderr:     at JobCollectionBase._DDPMethod_jobFail (packages/vsivsi:job-collection/shared.coffee:992:5)
App 402 stderr:     at Function.<anonymous> (packages/vsivsi:job-collection/server.coffee:98:20)
App 402 stderr:     at Function.<anonymous> (packages/vsivsi:job-collection/server.coffee:54:43)
App 402 stderr:     at methodCall (packages/vsivsi:job-collection/job/src/job_class.coffee:17:22)
App 402 stderr:     at Job.fail (packages/vsivsi:job-collection/job/src/job_class.coffee:752:14)
App 402 stderr:     at app/server/files/publications.js:209:21
App 402 stderr:     at runWithEnvironment (packages/meteor/dynamics_nodejs.js:108:1)
```

Also get following errors sometimes:

```
App 4751 stdout: Tue May 12 2015 18:24:24 GMT+0800 (CST), [SERVER], jobProgress, params: [{"_str":"5ab150e06c88899939aa8e6b"},{"_str":"9d3e4e6449a8b7f1875ea7f6"},20,100,{}]
App 4751 stdout: Tue May 12 2015 18:24:24 GMT+0800 (CST), [SERVER], jobProgress, params: [{"_str":"c5a6947fbc2bd0bba1264927"},{"_str":"9d3e4e6449a8b7f1875ea7f6"},20,100,{}]
App 4751 stdout: Tue May 12 2015 18:24:24 GMT+0800 (CST), [SERVER], jobProgress, returned: true
App 4751 stdout: Tue May 12 2015 18:24:24 GMT+0800 (CST), [SERVER], jobProgress, returned: true
App 4751 stdout: Exception in callback of async function: Error: gm().stream() or gm().write() with a non-readable stream.
App 4751 stdout:     at gm._spawn (/home/3dker/web/programs/server/npm/npm-container/node_modules/gm/lib/command.js:236:15)
App 4751 stdout:     at /home/3dker/web/programs/server/npm/npm-container/node_modules/gm/lib/command.js:148:19
App 4751 stdout:     at series (/home/3dker/web/programs/server/npm/npm-container/node_modules/gm/node_modules/array-series/index.js:11:36)
App 4751 stdout:     at gm._preprocess (/home/3dker/web/programs/server/npm/npm-container/node_modules/gm/lib/command.js:185:5)
App 4751 stdout:     at gm.stream (/home/3dker/web/programs/server/npm/npm-container/node_modules/gm/lib/command.js:146:10)
App 4751 stdout:     at app/server/files/publications.js:241:14
App 4751 stdout:     at runWithEnvironment (packages/meteor/dynamics_nodejs.js:108:1)
App 4751 stdout: Tue May 12 2015 18:24:24 GMT+0800 (CST), [SERVER], jobProgress, params: [{"_str":"c5a6947fbc2bd0bba1264927"},{"_str":"9d3e4e6449a8b7f1875ea7f6"},80,100,{}]
App 4751 stdout: Tue May 12 2015 18:24:24 GMT+0800 (CST), [SERVER], jobProgress, returned: true
App 4751 stdout: Tue May 12 2015 18:24:24 GMT+0800 (CST), [SERVER], jobDone, params: [{"_str":"c5a6947fbc2bd0bba1264927"},{"_str":"9d3e4e6449a8b7f1875ea7f6"},{},{}]
App 4751 stdout: Tue May 12 2015 18:24:24 GMT+0800 (CST), [SERVER], jobDone, returned: true
App 4751 stdout: Tue May 12 2015 18:24:24 GMT+0800 (CST), [SERVER], getWork, params: [["makeThumb"],{"maxJobs":1099}]
App 4751 stdout: Tue May 12 2015 18:24:24 GMT+0800 (CST), [SERVER], getWork, returned: []
App 4751 stdout: Tue May 12 2015 18:25:49 GMT+0800 (CST), [SERVER], jobProgress, params: [{"_str":"5ab150e06c88899939aa8e6b"},{"_str":"9d3e4e6449a8b7f1875ea7f6"},80,100,{}]
App 4751 stdout: Tue May 12 2015 18:25:49 GMT+0800 (CST), [SERVER], jobProgress, returned: true
App 4751 stdout: Tue May 12 2015 18:25:49 GMT+0800 (CST), [SERVER], jobDone, params: [{"_str":"5ab150e06c88899939aa8e6b"},{"_str":"9d3e4e6449a8b7f1875ea7f6"},{},{}]
App 4751 stderr: Warning: gridfs-locking-stream Write Lock Expired for file 23ca637ca86537806f72d997
App 4751 stdout: Tue May 12 2015 18:25:49 GMT+0800 (CST), [SERVER], jobProgress, params: [{"_str":"5ab150e06c88899939aa8e6b"},{"_str":"9d3e4e6449a8b7f1875ea7f6"},80,100,{}]
App 4751 stderr: Warning: gridfs-locking-stream Read Lock Expired for file b9ed18e88fce65907928b343
App 4751 stdout: Tue May 12 2015 18:25:49 GMT+0800 (CST), [SERVER], jobProgress, returned: true
App 4751 stdout: Tue May 12 2015 18:25:49 GMT+0800 (CST), [SERVER], jobDone, returned: true
App 4751 stdout: Tue May 12 2015 18:25:49 GMT+0800 (CST), [SERVER], jobDone, params: [{"_str":"5ab150e06c88899939aa8e6b"},{"_str":"9d3e4e6449a8b7f1875ea7f6"},{},{}]
App 4751 stderr: Running job not found { [String: '5ab150e06c88899939aa8e6b'] _str: '5ab150e06c88899939aa8e6b' } { [String: '9d3e4e6449a8b7f1875ea7f6'] _str: '9d3e4e6449a8b7f1875ea7f6' }
App 4751 stdout: Tue May 12 2015 18:25:49 GMT+0800 (CST), [SERVER], jobDone, returned: false
App 4751 stderr: Callback called multiple times in JobQueue
App 4751 stderr: /home/3dker/web/programs/server/node_modules/fibers/future.js:245
App 4751 stderr:                        throw(ex);
App 4751 stderr:                              ^
App 4751 stderr: Error: Callback was already called.
App 4751 stderr:     at packages/vsivsi:job-collection/job/src/job_class.coffee:136:19
App 4751 stderr:     at app/server/files/publications.js:221:13
App 4751 stderr:     at runWithEnvironment (packages/meteor/dynamics_nodejs.js:108:1)
```

what's wrong for me?

All ok under develop mode.

thx a lot 
victor

**Comment by xurwxj** [comment]
develop under osx 10.10.3
production under ubuntu 14.04

production mode also is normal under osx.

**Comment by vsivsi** [comment]
Thanks for reporting, I'm working on reproducing this issue.

**Comment by vsivsi** [comment]
Hi, I've been pounding on this for about an hour and have not been able to reproduce either of the errors you show above.  I'm using Ubuntu 14.04 server edition and deploying the most recent commit of the sample app using [mup](https://www.npmjs.com/package/mup).  

What commit of this package are you using?

Also, are these errors new since file-collection 1.1.0 was released?

Any more detailed steps you can give me to reproduce would be helpful... Are you uploading many image files at once?  What kind of images?  What version of graphicsmagick is installed?  Etc.

Please be as detailed and specific as you can.

**Comment by xurwxj** [comment]
happens after 1.1.0 released.
just upload one png image.

```
apt-cache show graphicsmagick
Package: graphicsmagick
Priority: optional
Section: universe/graphics
Installed-Size: 4410
Maintainer: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
Original-Maintainer: Daniel Kobras <kobras@debian.org>
Architecture: amd64
Version: 1.3.18-1ubuntu3
```

deploy order:
- meteor build command
  - svn commit for build result: bundle
  - svn update
  - cd programs/server
  - npm install
  - restart nginx with passenger
  - upload a png image

**Comment by vsivsi** [comment]
Hi, please try the latest commit and let me know if it either fixes the issue or gives more helpful error messages:  d1c2548548

**Comment by xurwxj** [comment]
oh, sorry for traveling, i'll try on next mondy.
thanks a lot.

**Comment by xurwxj** [comment]
under ubuntu env, now no error showed, but still get two job and generate two thumbs, sames "addedFileJob" was called twice.
everything is normal for mac env.

**Comment by xurwxj** [comment]
```
addedFileJob = function(file) {
        //  # Don't make new jobs for files tha already have them
        if (file && imageTypes[file.contentType] && !file.metadata._Job) {
                console.log(file);
```

```
App 13624 stdout: Thu May 21 2015 16:36:15 GMT+0800 (CST), [SERVER], getWork, returned: []
=====================================
App 13581 stdout: { length: 180523,
App 13581 stdout:   md5: 'c045d0f7abe0e467d3b579e6aaa1dde8',
App 13581 stdout:   uploadDate: Thu May 21 2015 16:36:34 GMT+0800 (CST),
App 13581 stdout:   chunkSize: 2096128,
App 13581 stdout:   filename: 'sss.png',
App 13581 stdout:   metadata: { owner: 'KiuCKYNmDjDPyvy6L' },
App 13581 stdout:   aliases: [],
App 13581 stdout:   contentType: 'image/png',
App 13581 stdout:   _id: { [String: '6da318bb8aa95b3cba70399f'] _str: '6da318bb8aa95b3cba70399f' } }
App 13624 stdout: { length: 180523,
App 13624 stdout:   md5: 'c045d0f7abe0e467d3b579e6aaa1dde8',
App 13624 stdout:   uploadDate: Thu May 21 2015 16:36:34 GMT+0800 (CST),
App 13624 stdout:   chunkSize: 2096128,
App 13624 stdout:   filename: 'sss.png',
App 13624 stdout:   metadata: { owner: 'KiuCKYNmDjDPyvy6L' },
App 13624 stdout:   aliases: [],
App 13624 stdout:   contentType: 'image/png',
App 13624 stdout:   _id: { [String: '6da318bb8aa95b3cba70399f'] _str: '6da318bb8aa95b3cba70399f' } }
========================================================
App 13581 stdout: Thu May 21 2015 16:36:36 GMT+0800 (CST), [SERVER], jobSave, params: [{"runId":null,"type":"makeThumb","data":{"owner":"KiuCKYNmDjDPyvy6L","inputFileURL":"http://test.3dker.com//gridfs/files/6da318bb8aa95b3cba70399f","outputFileURL":"http://test.3dker.com//gridfs/files/put/8bf3e9d9e4fd305a45c2ae3c","inputFileId":{"_str":"6da318bb8aa95b3cba70399f"},"outputFileId":{"_str":"8bf3e9d9e4fd305a45c2ae3c"}},"status":"waiting","updated":"2015-05-21T08:36:36.009Z","created":"2015-05-21T08:36:36.009Z","priority":0,"retries":6,"retryWait":20000,"retried":0,"retryBackoff":"constant","retryUntil":"+275760-09-13T00:00:00.000Z","repeats":0,"repeatWait":300000,"repeated":0,"repeatUntil":"+275760-09-13T00:00:00.000Z","after":"2015-05-21T08:36:41.010Z","progress":{"completed":0,"total":1,"percent":0},"depends":[],"resolved":[],"log":[{"time":"2015-05-21T08:36:36.010Z","runId":null,"level":"info","message":"Constructed"}]},{}]
App 13624 stdout: Thu May 21 2015 16:36:36 GMT+0800 (CST), [SERVER], jobSave, params: [{"runId":null,"type":"makeThumb","data":{"owner":"KiuCKYNmDjDPyvy6L","inputFileURL":"http://test.3dker.com//gridfs/files/6da318bb8aa95b3cba70399f","outputFileURL":"http://test.3dker.com//gridfs/files/put/a7df8ab05f858d48fcfcdaae","inputFileId":{"_str":"6da318bb8aa95b3cba70399f"},"outputFileId":{"_str":"a7df8ab05f858d48fcfcdaae"}},"status":"waiting","updated":"2015-05-21T08:36:36.013Z","created":"2015-05-21T08:36:36.013Z","priority":0,"retries":6,"retryWait":20000,"retried":0,"retryBackoff":"constant","retryUntil":"+275760-09-13T00:00:00.000Z","repeats":0,"repeatWait":300000,"repeated":0,"repeatUntil":"+275760-09-13T00:00:00.000Z","after":"2015-05-21T08:36:41.015Z","progress":{"completed":0,"total":1,"percent":0},"depends":[],"resolved":[],"log":[{"time":"2015-05-21T08:36:36.015Z","runId":null,"level":"info","message":"Constructed"}]},{}]
App 13624 stdout: Thu May 21 2015 16:36:36 GMT+0800 (CST), [SERVER], jobSave, returned: {"_str":"2330ad7d39fa8ee51ada5260"}
App 13581 stdout: Thu May 21 2015 16:36:36 GMT+0800 (CST), [SERVER], jobSave, returned: {"_str":"b43174f1b35b1b210fa081a4"}
App 13624 stdout: Thu May 21 2015 16:36:42 GMT+0800 (CST), [SERVER], getWork, params: [["makeThumb"],{"maxJobs":52}]
App 13581 stdout: Thu May 21 2015 16:36:42 GMT+0800 (CST), [SERVER], getWork, params: [["makeThumb"],{"maxJobs":52}]
App 13581 stdout: Thu May 21 2015 16:36:42 GMT+0800 (CST), [SERVER], getWork, returned: []
App 13624 stdout: Thu May 21 2015 16:36:42 GMT+0800 (CST), [SERVER], getWork, returned: [{"_id":{"_str":"2330ad7d39fa8ee51ada5260"},"runId":{"_str":"d94316241989aa8f5c802664"},"type":"makeThumb","data":{"owner":"KiuCKYNmDjDPyvy6L","inputFileURL":"http://test.3dker.com//gridfs/files/6da318bb8aa95b3cba70399f","outputFileURL":"http://test.3dker.com//gridfs/files/put/a7df8ab05f858d48fcfcdaae","inputFileId":{"_str":"6da318bb8aa95b3cba70399f"},"outputFileId":{"_str":"a7df8ab05f858d48fcfcdaae"}},"status":"running","updated":"2015-05-21T08:36:42.018Z","created":"2015-05-21T08:36:36.019Z","priority":0,"retries":5,"retryWait":20000,"retried":1,"retryBackoff":"constant","retryUntil":"+275760-09-13T00:00:00.000Z","repeats":0,"repeatWait":300000,"repeated":0,"repeatUntil":"+275760-09-13T00:00:00.000Z","after":"2015-05-21T08:36:41.015Z","progress":{"completed":0,"total":1,"percent":0},"depends":[],"resolved":[]},{"_id":{"_str":"b43174f1b35b1b210fa081a4"},"runId":{"_str":"d94316241989aa8f5c802664"},"type":"makeThumb","data":{"owner":"KiuCKYNmDjDPyvy6L","inputFileURL":"http://test.3dker.com//gridfs/files/6da318bb8aa95b3cba70399f","outputFileURL":"http://test.3dker.com//gridfs/files/put/8bf3e9d9e4fd305a45c2ae3c","inputFileId":{"_str":"6da318bb8aa95b3cba70399f"},"outputFileId":{"_str":"8bf3e9d9e4fd305a45c2ae3c"}},"status":"running","updated":"2015-05-21T08:36:42.018Z","created":"2015-05-21T08:36:36.013Z","priority":0,"retries":5,"retryWait":20000,"retried":1,"retryBackoff":"constant","retryUntil":"+275760-09-13T00:00:00.000Z","repeats":0,"repeatWait":300000,"repeated":0,"repeatUntil":"+275760-09-13T00:00:00.000Z","after":"2015-05-21T08:36:41.010Z","progress":{"completed":0,"total":1,"percent":0},"depends":[],"resolved":[]}]
App 13624 stdout: Thu May 21 2015 16:36:42 GMT+0800 (CST), [SERVER], jobProgress, params: [{"_str":"2330ad7d39fa8ee51ada5260"},{"_str":"d94316241989aa8f5c802664"},20,100,{}]
App 13624 stdout: Thu May 21 2015 16:36:42 GMT+0800 (CST), [SERVER], jobProgress, params: [{"_str":"b43174f1b35b1b210fa081a4"},{"_str":"d94316241989aa8f5c802664"},20,100,{}]
App 13624 stdout: Thu May 21 2015 16:36:42 GMT+0800 (CST), [SERVER], jobProgress, returned: true
App 13624 stdout: Thu May 21 2015 16:36:42 GMT+0800 (CST), [SERVER], jobProgress, returned: true
App 13624 stdout: Thu May 21 2015 16:36:42 GMT+0800 (CST), [SERVER], jobProgress, params: [{"_str":"b43174f1b35b1b210fa081a4"},{"_str":"d94316241989aa8f5c802664"},80,100,{}]
App 13624 stdout: Thu May 21 2015 16:36:42 GMT+0800 (CST), [SERVER], jobProgress, returned: true
App 13624 stdout: Thu May 21 2015 16:36:42 GMT+0800 (CST), [SERVER], jobDone, params: [{"_str":"b43174f1b35b1b210fa081a4"},{"_str":"d94316241989aa8f5c802664"},{},{}]
App 13624 stdout: Thu May 21 2015 16:36:42 GMT+0800 (CST), [SERVER], jobProgress, params: [{"_str":"2330ad7d39fa8ee51ada5260"},{"_str":"d94316241989aa8f5c802664"},80,100,{}]
App 13624 stdout: Thu May 21 2015 16:36:42 GMT+0800 (CST), [SERVER], jobProgress, returned: true
App 13624 stdout: Thu May 21 2015 16:36:42 GMT+0800 (CST), [SERVER], jobDone, params: [{"_str":"2330ad7d39fa8ee51ada5260"},{"_str":"d94316241989aa8f5c802664"},{},{}]
App 13624 stdout: Thu May 21 2015 16:36:42 GMT+0800 (CST), [SERVER], jobDone, returned: true
App 13624 stdout: Thu May 21 2015 16:36:42 GMT+0800 (CST), [SERVER], getWork, params: [["makeThumb"],{"maxJobs":51}]
App 13624 stdout: Thu May 21 2015 16:36:42 GMT+0800 (CST), [SERVER], jobDone, returned: true
App 13624 stdout: Thu May 21 2015 16:36:42 GMT+0800 (CST), [SERVER], getWork, returned: []
App 13624 stdout: Thu May 21 2015 16:36:42 GMT+0800 (CST), [SERVER], getWork, params: [["makeThumb"],{"maxJobs":52}]
App 13624 stdout: Thu May 21 2015 16:36:42 GMT+0800 (CST), [SERVER], getWork, returned: []
```

**Comment by vsivsi** [comment]
Okay thanks, I'll try this again on Ubuntu tomorrow...

**Comment by xurwxj** [comment]
difference between mac and ubuntu as following:
meteor run with 1 thread under mac when 2 thread through passenger under ubuntu.

so we should first update origin file obj when add to job.

```
addedFileJob = function(file) {
    //  # Don't make new jobs for files tha already have them
       // TODO update file obj first here to prevent undefined file.metadata._Job
    if (file && imageTypes[file.contentType] && !file.metadata._Job) {
```

**Comment by xurwxj** [comment]
i've set a flag with matteodem:server-session on the top of addFileJob function:

```
if (file && imageTypes[file.contentType] && !file.metadata._Job) {
        var file_id = file._id.toHexString();
        var file_s_s_id = ServerSession.get('thumb_f_' + file_id) || '';
        if (!file_s_s_id) {
            ServerSession.set('thumb_f_' + file_id, true);
```

problem solved through above func.

**Comment by xurwxj** [comment]
oh, just reduce the incidence through above func. in cluster env, big possible that several server threads will generate several jobs.
we need lock job in cluster env.
cluster env with unique mongodb connection.
Originally I hope to make a unified check through mongodb in the above func.

**Comment by vsivsi** [comment]
I see, so this isn't related to OSX vs Linux? It was a result of deploying to multiple Meteor servers running the same code but sharing a MongoDB instance. So each is responding to the `'added'` callback on the observe by creating a job?  I thought you said that it would even happen with a single server, which has me wondering why 'added' is being called twice for the same file on that observe?

I'll fix this, but honestly this app is meant to be a simple demonstration app, not a fully featured and multi-server scalable turnkey photo gallery app!

**Comment by vsivsi** [comment]
FYI, I've just tested this again on a single Ubuntu 14.04 instance with Meteor 1.1.0.2 in dev mode and I don't see any double job/thumbnail issue.  I'll try deploying it instead via Meteor-Up and see what I get.

**Comment by vsivsi** [comment]
I also see no problem on a headless Ubuntu 14.04 LTS server with the application built and deployed using Meteor-up:  https://github.com/arunoda/meteor-up

I understand what the code you included above does and why it fixes the problem you are seeing.  What I don't understand is how/why `addFileJob` is being called more than once per uploaded file in your environment.  Is it because you have multiple servers running that code against the same MongoDB instance?

**Comment by xurwxj** [comment]
the serve env:

```
ubuntu 14.04
meteor deployed under Phusion Passenger with 2 instance.
1 shared mongodb instance.
```

the problem will disappear if phusion passenger run with 1 instance.
actually, you are right no problem on any os with 1 meteor instance.

**Comment by vsivsi** [comment]
Okay, that makes sense now.  So your server-session solution doesn't work either because that state isn't shared between servers...

**Comment by xurwxj** [comment]
server-session solution do not work should be caused by concurrency.

**Comment by vsivsi** [comment]
Please try: https://github.com/vsivsi/meteor-file-job-sample-app/commit/96195675422135016cbf0156b070055f55140d66

A simple Mongo update can be used as a mutex

**Comment by xurwxj** [comment]
still exist, just reduce the incidence

**Comment by vsivsi** [comment]
Right. `update()` in Meteor isn't atomic.  I've committed a change to the dev branch that uses the atomic MongoDB `findAndModify()` method on the raw mongo collection.  It seems to be working here: https://github.com/vsivsi/meteor-file-job-sample-app/commit/f3fdb25f41b7baee1f12796cb7a6abffdb4b3df1

Honestly, I did this for you because I said I would, but it is way outside the scope of this simple demo app.  Dipping into the `rawCollection()` driver to atomically synchronize multiple servers is a pain and obfuscates the code, so I'm not sure I'm going to merge this into `master`.

Lesson: Scaling is non-trivial, and this has nothing to do with file-collection or job-collection, _per se_. They are each working as designed. The issue here is that my cute throwaway demo app failed to account for more complicated multi-server deployment scenarios, which I'm okay with.
